snap/snap.go

```diff
@@ -62,19 +63,109 @@ func addPointsAndSnap(ix *PointIndex, polygon *geom.Polygon) *geom.Polygon {
 				newPolygon = append(newPolygon, newRing)
 			}
 		default:
-			// deduplicate points in the ring, then add to the polygon
-			newPolygon = append(newPolygon, kmpDeduplicate(newRing))
+			// deduplicate points in the ring, check winding order, then add to the polygon
+			deduplicatedRing := kmpDeduplicate(newRing)
+			// inner rings (ringIdx != 0) should be clockwise
+			shouldBeClockwise := ringIdx != 0
+			// winding order is reversed if incorrect
+			validateWindingOrder(deduplicatedRing, shouldBeClockwise)
+			newPolygon = append(newPolygon, deduplicatedRing)
 		}
 	}
 	return (*geom.Polygon)(&newPolygon)
 }
 
+// validate winding order (CCW for outer rings, CW for inner rings)
+// if winding order is incorrect, ring is reversed to correct winding order
+func validateWindingOrder(ring [][2]float64, shouldBeClockwise bool) {
+	steps := []string{}
+	stepCounts := map[bool]int{true: 0, false: 0}
+	for i := 0; i < len(ring); i++ {
+		steps = append(steps, directionOfStep(ring[i], ring[(i+1)%len(ring)]))
+		// need at least a pair of steps to check winding order
+		if len(steps) <= 1 {
+			continue
+		}
+		// steps back (e.g., 'R,L') or repeated steps (e.g., 'R,R') count as a step for the winding order the ring should have
+		if steps[len(steps)-2] == oppositeDirectionOf(steps[len(steps)-1]) || steps[len(steps)-2] == steps[len(steps)-1] {
+			stepCounts[shouldBeClockwise]++
+		} else {
+			stepCounts[isClockwise(steps[len(steps)-2:])]++
+		}
+	}
+	if stepCounts[shouldBeClockwise] < stepCounts[!shouldBeClockwise] {
+		slices.Reverse(ring)
+	}
+}
+
+// returns if pair of steps is clockwise
+func isClockwise(stepsPair []string) bool {
+	clockwiseSteps := []string{
+		"U,R",   // up, then right
```

Comment:
misschien nog een enum (of golangs equivalent) hiervan maken. maar niet per se hoor

Reply:
Ik heb even zitten zoeken, maar ik vind niet direct een veel elegantere manier om dit te doen.